libp2p/protocols/pubsub/pubsubpeer.nim

```diff
@@ -330,6 +364,45 @@ proc canAskIWant*(p: PubSubPeer, msgId: MessageId): bool =
       return true
   return false
 
+proc clearSendPriorityQueue(p: PubSubPeer) =
+  while p.rpcmessagequeue.sendPriorityQueue.len > 0 and p.rpcmessagequeue.sendPriorityQueue[0].finished:
+    when defined(libp2p_expensive_metrics):
+      libp2p_gossipsub_priority_queue_size.dec(labelValues = [$p.peerId])
+    discard p.rpcmessagequeue.sendPriorityQueue.popFirst()
+
+proc sendNonPriorityTask(p: PubSubPeer) {.async.} =
+  while true:
+     while p.rpcmessagequeue.sendPriorityQueue.len > 0:
+       await p.rpcmessagequeue.sendPriorityQueue[0]
+       p.clearSendPriorityQueue()
+     # we send non-priority messages only if there are no pending priority messages
+     let msg = await p.rpcmessagequeue.nonPriorityQueue.popFirst()
```

Comment:
while waiting here, a priority message might have been added - this await should be at the top of the while

Reply:
But then we won't clear the priority queue if there's no non-priority msg. Actually, it's true for both solutions.

Reply:
the prio queue can be cleared before adding to it (also)